### PR TITLE
Make sure logicalId is unique in test suite

### PR DESCRIPTION
**Pull request by kroesctrl**
Signed-off-by: Gerben Kroes <gerben@kroesctrl.nl>

### CODE REVIEW

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/database/device/SimulatorDeviceRegister.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.Map;
+import java.util.Optional;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+@Slf4j
+@Component
+public class SimulatorDeviceRegister {
+  private final HashMap<Long, Map<String, Long>> logicalIdByPortByDeviceIdentification =
```
Comment (harrymiddelburgetall) [inline integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/database/device/SimulatorDeviceRegister.java:23]: private final **Map**<Long, Map<String, Long>> logicalIdByPortByDeviceIdentification =
      new HashMap<>();

**Review by harrymiddelburgetall — CHANGES_REQUESTED** [review]
Congratulations! Nice improvement of the tests with simulators. Just a few remarks.

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/database/device/SimulatorDeviceRegister.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+          "Found existing logicalDeviceId: {}, device: {}, port: {}",
+          logicalId,
+          deviceIdentification,
+          port);
+      return logicalId;
+    }
+    final Optional<Long> lastLogicalId =
+        this.logicalIdByPortByDeviceIdentification.get(port).values().stream().max(Long::compareTo);
+
+    final Long newLogicalId = lastLogicalId.orElse(0L) + 1;
```
[the remark this answers is not in the document]
Comment (kroesctrl) [reply integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/database/device/SimulatorDeviceRegister.java:51]: The last logicalId should be calculated by port

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/database/device/SimulatorDeviceRegister.java (hunk 1 above)
Comment (kroesctrl) [reply integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/database/device/SimulatorDeviceRegister.java:23]: Chcek, aangepast

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/database/device/DlmsDeviceSteps.java [hunk 3]
```diff
@@ -161,7 +163,23 @@ public void aDlmsDevice(final Map<String, String> inputSettings) {
 
     this.createDeviceAuthorisationInCoreDatabase(device);
 
-    this.createDlmsDeviceInProtocolAdapterDatabase(inputSettings);
+    final DlmsDevice dlmsDevice = this.createDlmsDeviceInProtocolAdapterDatabase(inputSettings);
+    this.setLogicalDeviceId(dlmsDevice);
```
[the remark this answers is not in the document]
Comment (kroesctrl) [reply integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/database/device/DlmsDeviceSteps.java:167]: Refactored

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/database/device/DlmsDeviceSteps.java [hunk 4]
```diff
@@ -161,7 +163,23 @@ public void aDlmsDevice(final Map<String, String> inputSettings) {
...
+   */
+  private void setLogicalDeviceId(final DlmsDevice dlmsDevice) {
+    if (dlmsDevice.getPort() == null || dlmsDevice.getDeviceIdentification() == null) {
+      return;
+    }
+    final Long logicalId =
+        this.simulatorDeviceRegister.getLogicalId(
+            dlmsDevice.getPort(), dlmsDevice.getDeviceIdentification());
+    dlmsDevice.setLogicalId(logicalId);
+    this.dlmsDeviceRepository.save(dlmsDevice);
```
[the remark this answers is not in the document]
Comment (kroesctrl) [reply integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/database/device/DlmsDeviceSteps.java:182]: Changed

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/database/device/SimulatorDeviceRegister.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (harrymiddelburgetall) [reply integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/database/device/SimulatorDeviceRegister.java:51]: true

**Review by harrymiddelburgetall — APPROVED** [review]

**Comment by kroesctrl** [comment]
https://ci.opensmartgridplatform.org/job/OSGP_open-smart-grid-platform_pr/3186/

**Comment by kroesctrl** [comment]
Retest this please

**Pull request merged**